### PR TITLE
Issue #2366: Infra for portfolio permissions

### LINKED ISSUE #2366: Update User object and add permission logic
### Issue description
For the org model MVP, we have several features that we need to show/hide based on user permissions, such as domain management, managing requests, user management (at the whole portfolio level) and so on. That said, as we build out views for domain management in particular, we are going to need to have a permission system in place so that we can get started with these conditional views. 

We also, are only showing portfolio based on the creator of the portfolio, but the goal will be to retrieve the portfolio that the user has permissions to, just like we currently get the domains they have management rights to. 

The third thing permissions will be used for is showing a human readable permission next to user names such as "org member", "domain manager" "admin" and "read-only admin". This human readable name will eventually need to appear   in the both django admin (for easy analyst viewing) and in a Portfolio User Management page to our users.

The scope of this ticket:
Given the above needs, set up the foundation for permissions, such that at the completion of this ticket, a developer could 

1. add a a role to their User account and turn on/off request access
2. call a function to check if their account has a given permission (such as would be easy to do in a view)

### Acceptance criteria

- [ ] User is updated so that a portfolio can be added to a User's account along with roles and additional access flag( note at this time it would just be request_access) [ the miro](https://miro.com/app/board/uXjVKAEDJB4=/?moveToWidget=3458764591533149228&cot=10) 
- [ ] the ground work is set so that future tickets can easily check if a feature is allowed for the given user. 
- [ ] a mapping is added that correlates permissions to roles/access flags etc to their correct permissions  [ see this permission table for specifics](https://miro.com/app/board/uXjVKAEDJB4=/?moveToWidget=3458764591825956560&cot=14)
- [ ] In the future, it will be easy to adjust which "permissions" go with each role and boolean "access flags" in the future via a code update.
- [ ] Django permissions on the user object are not used 

### Additional context
Note I am calling the boolean flags for adjustable "permissions" on a user to access flags in this ticket, purely for lack of better terminology for it.

Implementation notes:
As long as we have a way to start accessing views based on role/permissions going forward. The implementation does not have to directly follow this but this is what is currently the thought here:

- [ ] instead of Django's built in permissions we store a mapping of role to a custom permission value
- [ ] functions are added that retrieve all "permissions" (for lack of a better term) for the user. This includes for their given role , their access flag ( only manage_request at the moment), and domain management abilities (based off of user domain roles). Then checks against the combined list of "permissions" to see if the user has the desired item.

For custom permissions, please avoid reusing hard coded strings and use enums where possible.

**additional comments for rachid and dave**
I suggest adding a mapping for domain manager access that checks if the user has user domain roles instead of for the specific permission mapping. That way we don't have one way for checking permissions on request and another for managing domains view. (ie both could check in the view just off the self.user object when in the org mode but we keep the domain management logic still dependent on having that role). 

### Links to other issues

- #2367

### PR DESCRIPTION
**Pull request by rachidatecs**
## Ticket

Resolves #2366

## Changes

- Portfolio roles and permissions added to User model.  User gets portfolio permissions associated with a portfolio role when the portfolio role is assigned.
- Portfolio additional permissions added to User model.  In addition to permissions assigned through roles, a user can be granted additional portfolio permissions.
- Portfolio roles and additional permissions can be managed through the User edit form in django admin
- Middleware in the application redirects user to appropriate landing page when organization_feature waffle is on and user is assigned a portfolio, based on the user's portfolio roles and additional permissions
- Portfolio views are protected by portfolio roles and additional permissions
- Portfolio menu options are based on portfolio roles and additional permissions

## Context for reviewers

Notes for dev review:
* Roles and permissions, and permission checks, are all defined in the User model
* Needed to add custom widget for managing an array of roles and permissions in django admin

Notes for design review:
Assign yourself a portfolio.  Make sure the waffle flag for organization_feature is ON.  Edit your user's portfolio roles and permissions in django admin.  Navigate to home page.  Verify that you land on the correct view, and that the menu items are appropriate for your role/permission level  

#### Satisfied acceptance criteria and met development standards

- [ ] Met the acceptance criteria, or will meet them in a subsequent PR
- [ ] Created/modified automated tests
- [ ] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [ ] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [ ] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

File: src/registrar/models/user.py
Comment (abroddrick) [inline src/registrar/models/user.py:87]: We do have some users  (admin and read only admins) who will be able to view all requests, but not edit the requests they did not make. However, these same people can make and edit requests.  On the flipside, others will follow our normal process where they just see requests they created and can manage those requests. 

With that said, should this be changed to include VIEW_ALL_REQUESTS? Then View_requests could be VIEW_CREATED_REQUESTS. Just to differentiate between the permissions?

File: src/registrar/models/user.py
Comment (abroddrick) [inline src/registrar/models/user.py:281]: @rachidatecs is this not ready for review?

File: src/registrar/models/user.py
Comment (abroddrick) [inline src/registrar/models/user.py:265]: **suggestion** "this guy" 👀. Maybe remove that. Also, are views then expected to pass in things like "view_permission" etc? if so I would suggest moving the permissions enum to a utility to avoid any chance of circular imports down the road with multiple files using those enums by importing from user.

Comment (rachidatecs) [reply src/registrar/models/user.py:265]: I like the idea of moving the enum to a utility but I don't get how we're risking circular imports (right now a view would import from user for both the constants and the function. The function refers to its class. A circular dependency would imply that something in user would need to import from the view, no?)

File: src/registrar/models/user.py
Comment (rachidatecs) [reply src/registrar/models/user.py:281]: Stray comment?

File: src/registrar/models/user.py
Comment (rachidatecs) [reply src/registrar/models/user.py:87]: Same thing on DOMAINS I think?

Comment (rachidatecs) [reply src/registrar/models/user.py:87]: VIEW_MANAGED_DOMAINS vs VIEW_ALL_DOMAINS

File: src/registrar/models/user.py
Comment (rachidatecs) [reply src/registrar/models/user.py:265]: In other words, isn't this a similar architecture to how we're handling domain state and request status, where these enums are a class prop and there are methods that we pass these to?

**Review by abroddrick — COMMENTED** [review]
When I followed the steps and 1.) turned on the flag and 2) assigned myself to a portfolio, I got a combination of the two views: 
image

Also on the User admin, can you change portfolio to be a combo box? Right now it's not super needed given we don't have any portfolios, but we will need it down the road I am sure

File: src/registrar/tests/test_reports.py
Comment (abroddrick) [inline src/registrar/tests/test_reports.py:562]: nice catch, that's ugly

File: src/registrar/tests/test_models.py
Comment (abroddrick) [inline src/registrar/tests/test_models.py:1222]: **suggestion:** These all really sound like different tests. do they all have to be in this one unit test as opposed to breaking them out?

File: src/registrar/tests/test_views_portfolio.py
Comment (abroddrick) [inline src/registrar/tests/test_views_portfolio.py:31]: test comment here please

File: src/registrar/tests/test_views_portfolio.py
Comment (abroddrick) [inline src/registrar/tests/test_views_portfolio.py:38]: love this comment!

File: src/registrar/tests/test_views_portfolio.py
Comment (abroddrick) [inline src/registrar/tests/test_views_portfolio.py:45]: code comment

File: src/registrar/tests/test_views_portfolio.py
Comment (abroddrick) [inline src/registrar/tests/test_views_portfolio.py:59]: comments here

File: src/registrar/tests/test_views_portfolio.py
Comment (abroddrick) [inline src/registrar/tests/test_views_portfolio.py:75]: comments here and on the rest of these unit tests. They are very helpful when tests start randomly failing

File: src/registrar/tests/test_views_portfolio.py
Comment (abroddrick) [inline src/registrar/tests/test_views_portfolio.py:146]: Lol, but also, change this line 😂

File: src/registrar/tests/test_views_portfolio.py
Comment (abroddrick) [inline src/registrar/tests/test_views_portfolio.py:170]: Some more inline comments here would be good. I wasn't expecting the test to continue based on the name, so maybe quick line "reducing permissions to basic view perms hides domains/domain request nav"

File: src/registrar/tests/test_views_portfolio.py
Comment (abroddrick) [inline src/registrar/tests/test_views_portfolio.py:107]: only one of these comments is needed, though I like the clarity here ;)

File: src/registrar/tests/test_views_portfolio.py
Comment (abroddrick) [inline src/registrar/tests/test_views_portfolio.py:124]: same duplicate line

File: src/registrar/tests/test_views_portfolio.py
Comment (abroddrick) [inline src/registrar/tests/test_views_portfolio.py:141]: duplicate here

File: src/registrar/views/utility/permission_views.py
Comment (abroddrick) [inline src/registrar/views/utility/permission_views.py:175]: great comment here

File: src/registrar/views/utility/permission_views.py
Comment (abroddrick) [inline src/registrar/views/utility/permission_views.py:191]: Add a quick comment here and below for the sake of new people to the code base backtracking things in their IDE

**Review by abroddrick — COMMENTED** [review]
Found out the issue I posted about an hour ago was that I didn't save the permission changes and had only saved adding the portfolio to the user. If the user doesn't have permissions added, should they even see any part of the template? I would think not, with us switching to be all permission based. Curious on your thoughts here. Also added a couple questions and minor suggestions/ code clean up items.

File: src/registrar/admin.py
Comment (dave-kennedy-ecs) [reply src/registrar/admin.py:110]: comment added

File: src/registrar/models/user.py
Comment (dave-kennedy-ecs) [reply src/registrar/models/user.py:273]: thank you, will clean up note in next ticket

File: src/registrar/models/user.py
Comment (dave-kennedy-ecs) [reply src/registrar/models/user.py:289]: this is the most basic permission shared by all types of portfolio users (members, admins, and admin read only).  it is the permission we are checking to determine basic layout of views, whether or not to display portfolio menus.  i think this should remain VIEW_PORTFOLIO.  do you think that has_base_portfolio_permissions is confusing as a name?

File: src/registrar/models/user.py
Comment (dave-kennedy-ecs) [reply src/registrar/models/user.py:298]: I think there has to be a general has_domain_requests_permission and a general has_domains_portfolio_permission.  These permission checks are used to route requests in middleware as well as make decisions on what to display in a view or if to display a view.  so i think these general permission checks need to exist and encompass all three domain portfolio permissions or all three domain request portfolio permissions.  I have therefore made a change to the logic of these three methods; hope it makes sense and is correct

**Comment by dave-kennedy-ecs** [comment]
> Found out the issue I posted about an hour ago was that I didn't save the permission changes and had only saved adding the portfolio to the user. If the user doesn't have permissions added, should they even see any part of the template? I would think not, with us switching to be all permission based. Curious on your thoughts here. Also added a couple questions and minor suggestions/ code clean up items.

I modified so the display of the template is based on permissions.  It was previously based on portfolio creator (which was from before we added portfolio permissions).

so, for the portfolio templates to show for a user, there must be three conditions satisfied now: the user must have organization_feature enabled (either individually or globally), the user must have a portfolio assigned, the user must have the base portfolio permission (VIEW_PORTFOLIO)

File: src/registrar/models/user.py
Comment (dave-kennedy-ecs) [reply src/registrar/models/user.py:298]: resolving conversation based on previous comment

File: src/registrar/views/utility/permission_views.py
Comment (dave-kennedy-ecs) [reply src/registrar/views/utility/permission_views.py]: done

File: src/registrar/tests/test_models.py
Comment (dave-kennedy-ecs) [reply src/registrar/tests/test_models.py:1222]: THey don't have to be in the same test.  Just did so for simplicity

File: src/registrar/tests/test_views_portfolio.py
Comment (dave-kennedy-ecs) [reply src/registrar/tests/test_views_portfolio.py:167]: done

File: src/registrar/models/user.py
Comment (dave-kennedy-ecs) [reply src/registrar/models/user.py]: i agree with rachid.  i think we are following de facto standard established in these other models

File: src/registrar/models/portfolio.py
Comment (dave-kennedy-ecs) [reply src/registrar/models/portfolio.py]: yes, i agree.  made the change to 'created_portfolios'

File: src/registrar/tests/test_views_portfolio.py
Comment (rachidatecs) [reply src/registrar/tests/test_views_portfolio.py]: You're breaking my heart

Comment (rachidatecs) [reply src/registrar/tests/test_views_portfolio.py]: But ok

File: src/registrar/admin.py
Comment (dave-kennedy-ecs) [reply src/registrar/admin.py:131]: this is done, rachid did so

File: src/registrar/forms/portfolio.py
Comment (abroddrick) [inline src/registrar/forms/portfolio.py:71]: This looks likes just a copy and past from DomainOrgNameAddressForm. I think it can be deleted. If there is another purpose here (such as for a later ticket) add a quick comment before merging

File: src/registrar/models/user.py
Comment (abroddrick) [inline src/registrar/models/user.py:67]: Lol, I'll allow it

File: src/registrar/models/user.py
Comment (abroddrick) [reply src/registrar/models/user.py:90]: Yes!

File: src/registrar/models/user.py
Comment (abroddrick) [reply src/registrar/models/user.py:288]: The base permission one (has base portfolio_permissions) does make sense! I was referring to the domains_portfolio permission which is tied to "view all domains", I do think we will eventually need to differentiate between view_all domains and viewing the domains tab in general as the naming here implies viewing just domains, when really it's the permission to see all domains. That said it can be handled in the follow on ticket and not something I am blocking on

File: src/registrar/models/user.py
Comment (abroddrick) [reply src/registrar/models/user.py:298]: I agree with the general check for sure, maybe it's actually the permission it gets that should be adjusted. For instance I would think both requestors, admins and read-only admins have a shared permission to see domain requests and then there is a permission to view all domain requests that ONLY admins/read-only admins get. The general "request" permission checkout need to check for the permission that all three groups share ie just viewing domain requests in general, whereas here the name of the permission is view all which seems to apply to admins and read-only admins. Again though, not blocking on this, when we get to the request pages we can adjust the permissions when/if needed

**Review by abroddrick — APPROVED** [review]
LGTM, no blocking comments, just responded to a few comments, but I am good with things as they are and we can always adjust as we get into various pages. Thank you both!!

File: src/registrar/forms/portfolio.py
Comment (dave-kennedy-ecs) [reply src/registrar/forms/portfolio.py]: removed

**Pull request merged**
